tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts

```diff
@@ -48,90 +46,97 @@ export class LinkedTimeFobControllerComponent {
   @Input() axisDirection!: AxisDirection;
-  @Input() steps!: number[];
   @Input() linkedTime!: LinkedTime;
-  @Input() temporalScale!: TemporalScale;
+  @Input() cardAdapter!: FobCardAdapter;
   @Output() onSelectTimeChanged = new EventEmitter<LinkedTime>();
 
   private currentDraggingFob: Fob = Fob.NONE;
 
-  private currentDraggingFobUpperBoundIndex: number = -1;
-  private currentDraggingFobLowerBoundIndex: number = -1;
-
   // Helper function to check enum in template.
   public FobType(): typeof Fob {
     return Fob;
   }
 
   getCssTranslatePx(step: number): string {
     if (this.axisDirection === AxisDirection.VERTICAL) {
-      return `translate(0px, ${this.temporalScale(step)}px)`;
+      return `translate(0px, ${this.cardAdapter.getAxisPositionFromStep(
+        step
+      )}px)`;
     }
 
-    return `translate(${this.temporalScale(step)}px, 0px)`;
+    return `translate(${this.cardAdapter.getAxisPositionFromStep(
+      step
+    )}px, 0px)`;
   }
 
   startDrag(fob: Fob) {
     this.currentDraggingFob = fob;
-    this.currentDraggingFobUpperBoundIndex =
-      this.getDraggingFobUpperBoundIndex();
-    this.currentDraggingFobLowerBoundIndex =
-      this.getDraggingFobLowerBoundIndex();
   }
 
   stopDrag() {
     this.currentDraggingFob = Fob.NONE;
-    this.currentDraggingFobUpperBoundIndex = -1;
-    this.currentDraggingFobLowerBoundIndex = -1;
   }
 
   mouseMove(event: MouseEvent) {
     if (this.currentDraggingFob === Fob.NONE) return;
 
-    let newLinkedTime = this.linkedTime;
-    if (this.isDraggingHigher(event.clientY, event.movementY)) {
-      const stepAbove =
-        this.steps[this.getStepIndexHigherThanMousePosition(event.clientY)];
-      if (this.currentDraggingFob === Fob.END) {
-        newLinkedTime.end!.step = stepAbove;
-      } else {
-        newLinkedTime.start.step = stepAbove;
-      }
-      this.onSelectTimeChanged.emit(newLinkedTime);
+    const newLinkedTime = this.linkedTime;
+    let newStep: number | null = null;
+    const mousePosition =
+      event.clientY -
+      this.axisOverlay.nativeElement.getBoundingClientRect().top;
+    if (this.isDraggingHigher(mousePosition, event.movementY)) {
+      newStep = this.cardAdapter.getStepHigherThanAxisPosition(mousePosition);
+    } else if (this.isDraggingLower(mousePosition, event.movementY)) {
+      newStep = this.cardAdapter.getStepLowerThanAxisPosition(mousePosition);
+    }
+
+    if (newStep === null) {
+      return;
     }
 
-    if (this.isDraggingLower(event.clientY, event.movementY)) {
-      const stepBelow =
-        this.steps[this.getStepIndexLowerThanMousePosition(event.clientY)];
-      if (this.currentDraggingFob === Fob.END) {
-        newLinkedTime.end!.step = stepBelow;
-      } else {
-        newLinkedTime.start.step = stepBelow;
+    if (this.currentDraggingFob === Fob.END) {
+      // Do not let the end fob pass the start fob.
+      // TODO: add swapping logic here to allow continued dragging
```

Comment:
It's better to have your username or bug number in a TODO
`TODO(b/...):` or `TODO(jieweiwu):..`